Calendar arithmetic and duration fixes

bobhy opened the pull request:
# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->
Fix Duration type so we can safely add back month and year duration constants. 
Address financial and statistical calculations that are currently hard.

Lord help me, I've written a whole spec!  It's in draft form as a file in the PR, to be removed before merging.
See README-duration-spec.md

In the meantime, reviewers can comment on and request changes via usual PR review mechanism.

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect -A clippy::result_large_err` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass
- `cargo run -- crates/nu-std/tests/run.nu` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->

bobhy commented:
expect this will also close #8355

fdncred commented:
@bobhy I missed that there is actual code here now vs markdown. How's this going? Your code looks interesting.

fdncred commented on crates/nu-protocol/src/value/duration.rs:
```diff
@@ -0,0 +1,239 @@
+    pub hours: UnitSize,
+    pub minutes: UnitSize,
+    pub seconds: UnitSize,
+    pub fraction: UnitSize, // fractional number of nanoseconds (< 1 sec)
```
The only thing that has made me question this, so far, is why we don't have milliseconds, microseconds, and nanoseconds but instead of fraction. 🤔

bobhy replied:
This is the utility module supporting the value, the parser will have constants for all the time units.

fdncred commented:
> Just checked it in last night. Making good progress, could have it running by end of week.

Great! Wasn't trying to pressure you or anything. I haven't been watching too closely. Glad you're making progress.

fdncred commented:
Looks like you're chugging along. Let me know when there's something to try/play with and I'll give some feedback.

Also, dropping these two things here for inspiration.
1. [when](https://github.com/mitsuhiko/when) - I thought this seemed nushell-y because of the natural human language aspects.
2. [dateparse](https://github.com/waltzofpearls/dateparser) - I liked seeing all the datetime parsing this crate could do. I'm not sure if it's better/worse/same as `dtparse`, which I think we use, but thought it was interesting.

bobhy commented:
Hmm, I've been spending a lot of quality time lately with `chrono` and can
see why there are so many other rust calendar packages in the library.  I
didn't look into dateparse, but dtparse was already pretty good, and the
developer responded quickly with an enhancement when we asked for it.  It's
too bad, kind of for the whole Rust ecosystem, that chrono is not maturing
faster.  Calendar arithmetic is tricky, but not as complicated as they seem
to think.

`what` is cool.  It clearly answers more calendar questions than any single
Nushell user would ever be likely to have.  The question is, how to package
it up for Nushell use?  Small script module wrapper around the existing
CLI?  direct built-in integration (seems overkill for the limited
audience)?  Plugin? (not clear what the niche is for plugins now that
script modules are maturing nicely).

Today in this PR, it's integrating new duration type into
`nu_protocol::value::` and all the math operations with duration.
Tomorrow, the parser and last, visit the bultin commands.  Still feeling
good!







Hmm, we do already use `dtparse` and developer was responsive in adding a
small enhancement for us when I asked for it.

On Wed, May 31, 2023 at 7:53 AM Darren Schroeder ***@***.***>
wrote:

> Looks like you're chugging along. Let me know when there's something to
> try/play with and I'll give some feedback.
>
> Also, dropping these two things here for inspiration.
>
>    1. when <https://github.com/mitsuhiko/when> - I thought this seemed
>    nushell-y because of the natural human language aspects.
>    2. dateparse <https://github.com/waltzofpearls/dateparser> - I liked
>    seeing all the datetime parsing this crate could do. I'm not sure if it's
>    better/worse/same as dtparse, which I think we use, but thought it was
>    interesting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nushell/nushell/pull/9113#issuecomment-1570048637>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPBIZCWFGFMLNDJHZHAQQTXI4WKXANCNFSM6AAAAAAXXT4IPA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

fdncred commented:
> `what` is cool.  It clearly answers more calendar questions than any single Nushell user would ever be likely to have

agreed. i've said many times to the core-team that chrono does the date math correctly. the chrono-humanize also spits it out kind of how we do, with slight differences.

dateparse is cool but wouldn't want to change from dtparse unless we needed to. just wanted to bring it to your attention in case we needed something different.

> The question is, how to package it up for Nushell use?

I'd like to see date/datetime/duration math in rust, but it would cool to hook into the power of chrono/dtparse for advanced scripting things.

One of the keys to making scripting/stdlib/modules more mature is the ability to have all the "building blocks" at your fingertips. We have some of those for datetime now but I'm not sure we have enough. I can envision more date commands, especially with how finicky timezones are like why is PDT not a timezone and chrono even recognizes it but refuses to use it? I mean who is really going to remember America/Los_Angeles when thinking PDT. But that's a different PR.

Example
```
"1pm on Monday PDT" | into datetime
tzname PDT identified but not understood.
Mon, 05 Jun 2023 13:00:00 -0500 (in 3 days)
```

bobhy commented:
New duration type exists and can do the essential calendar operations.  
It's plumbed into the parser and into protocol.
It runs `cargo test -p nu_protocol` and `-p nu_parser`, though the built-in commands don't even compile yet, they need to track some name changes.
Left to do:
* convert the builtins and random other dependencies on the old duration.
* implement specialized `date diff`,  `into duration` and `into record`.
* unit tests for nu_protocol (math operations), and for builtins.

fdncred commented:
@bobhy iirc, you're relatively new to rust. So, I wanted to share something I learned a while back that could be a fit in these type of changes. Don't feel like you have to use this technique. I'm just sharing a rust "trick".

Rust has these things called extension traits which allows you to add functions to any existing type that you don't have the source code too, like chrono.

Here's a neat example of how the author is extending `chrono::NaiveDate` so they can just call their new functions on any naivedate variable.

https://github.com/b1rger/carl/blob/main/src/lib/date_extensions.rs

bobhy commented:
This build is worth a light try. 
Basic duration arithmetic is working: `1_day + 22_hr`, most of the `duration | to ...` and `into` commands work, even `list<duration> | math avg`
`sleep` is broken, `math mode`, too.  from and to duration don't work at all.  and `from nuon` and `from json` will need work too.  New serialize format for durations has changed, we'll have to do extra work to read, e.g, a file you saved last week that contains string durations (as i64s), vs new format `3_day`.

fdncred commented:
Just playing around.
```
❯ 1day + 22hr
46_hours
❯ 46hrs
46_hours
❯ 1day + 22hr + 1wk
214_hours
```
I'm not sure I like the default return with a `_` separating number & unit. I think it should remain as it was before, without the underscore.
How is it determining the unit when doing math? I'm trying to understand why `1day + 22hr + 1wk` equals `214hrs` instead of `1wk 1day 22hrs`.

fdncred commented:
This is kind of cool too
```
❯ 1day + 22hr + 1wk | debug -r
Duration {
    val: NuDuration {
        quantity: 214,
        unit: Hour,
    },
    span: Span {
        start: 307398,
        end: 307415,
    },
}
❯ 1day + 22hr + 1wk | into record
╭──────────┬───────╮
│ quantity │ 214   │
│ unit     │ hours │
╰──────────┴───────╯
```

Weird error
```
[14_days 12_years 0_ns 12_years, 10_mos] | math mode
Error: nu::shell::unimplemented

  × Function not implemented revers1ble hash for Duration
   ╭─[entry #12:1:1]
 1 │ [14_days 12_years 0_ns 12_years, 10_mos] | math mode
   ·                                            ────┬────
   ·                                                ╰── revers1ble hash for Duration
   ╰────
```

bobhy commented:
In 'duration1` + `duration2` the one with the larger time unit is scaled to
the smaller so the result doesn't lose precision. There will be an
`duration | into duration -- units <duration>` so you can specify the units
(and lose precision) directly.

I *like* the underscores.  Maybe there should be a preference item to
contron le default format, like for filesize?

`math mode` is a problem.  It depends on there being an 8 byte binary repr
for each input which is **reversible**.  Sadly, NuDuration needs 12 bytes:
for quantity **and** units.  But there are other ways to do the calc
without depending on a reversible hash, that would simplify math mode a
lot.  For now, I'm simply removing support for duration from the command,
will fix it in another pr.


On Sun, Jun 4, 2023, 11:15 PM Darren Schroeder ***@***.***>
wrote:

> This is kind of cool too
>
> ❯ 1day + 22hr + 1wk | debug -r
> Duration {
>     val: NuDuration {
>         quantity: 214,
>         unit: Hour,
>     },
>     span: Span {
>         start: 307398,
>         end: 307415,
>     },
> }
> ❯ 1day + 22hr + 1wk | into record
> ╭──────────┬───────╮
> │ quantity │ 214   │
> │ unit     │ hours │
> ╰──────────┴───────╯
>
> Weird error
>
> [14_days 12_years 0_ns 12_years, 10_mos] | math mode
> Error: nu::shell::unimplemented
>
>   × Function not implemented revers1ble hash for Duration
>    ╭─[entry #12:1:1]
>  1 │ [14_days 12_years 0_ns 12_years, 10_mos] | math mode
>    ·                                            ────┬────
>    ·                                                ╰── revers1ble hash for Duration
>    ╰────
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nushell/nushell/pull/9113#issuecomment-1575978341>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPBIZCK4HVPUHAB4GQJVTTXJVFLVANCNFSM6AAAAAAXXT4IPA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

fdncred commented:
I don't understand what you mean by loser precision with something like this.
```
> 1day + 22hr
46hrs
```
vs
```
> 1day + 22hr
1day 22hrs
```
Also, we've discussed this before that this type of math needs to spit out the same way it does now but more accurate if possible, with weeks and months.
```
> (date now) - 2019-05-10T09:59:12-07:00
4yrs 24days 20hrs 18mins 15secs 754ms 840µs
```

bobhy commented:
E.g adding 1_day and 22_hours: you can't add directly, so you have to
convert them to some common unit of measure.  If I converted 22 hours to
days, it would have to be truncated to 0 days.  If I converted both to
nanoseconds, I might get an overflow (these quantities are int).  So
instead, I convert 1 day (the larger unit) to 24 hours (the smaller unit)
and can then add to 22 hours without losing precision.

The key point of the new duration type is that it's a quantity and a unit
of measure in a struct, not just a quantity of nanoseconds under the
covers, so it can represent months or years accurately.  And the quantity
is an int, not a float, to avoid accumulating roundoff error.

When you try to add days to months, you can only do it if user provides a
specific date to start counting from (so you know how many days are in that
particular month, etc).  In an ordinary arithmetic expression, `<duration>
+ <duration>`, there's no place to provide the base date, so I have to give
an error if the durations aren't in the same range of units (either both
ns...weeks or both months ...) .  But there is a specialized command that
lets you specify all 3 inputs and get an updated duration.


On Mon, Jun 5, 2023 at 9:50 AM Darren Schroeder ***@***.***>
wrote:

> I don't understand what you mean by loser precision with something like
> this.
>
> > 1day + 22hr
> 46hrs
>
> vs
>
> > 1day + 22hr
> 1day 22hrs
>
> Also, we've discussed this before that this type of math needs to spit out
> the same way it does now but more accurate if possible, with weeks and
> months.
>
> > (date now) - 2019-05-10T09:59:12-07:00
> 4yrs 24days 20hrs 18mins 15secs 754ms 840µs
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nushell/nushell/pull/9113#issuecomment-1576845363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPBIZCE4IACUEMTZG3EMF3XJXP2XANCNFSM6AAAAAAXXT4IPA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

fdncred commented:
> adding 1_day and 22_hours: you can't add directly ... and can then add to 22 hours without losing precision.

right. this all makes sense to me and sounds correct. what i'm saying doesn't look right is how a duration is represented on output. Some of these may be easy while others may not work like adding days to months unless a given start date is provided. I'm not sure how we make that more intuitive. I'm not a fan of always spitting out the smallest unit available. Even if it's more accurate, it may seem like a regression.

This should work though eventually since they're both the same data type. `(date now) - 2019-05-10T09:59:12-07:00` Right now it just shows a bunch of unhelpful nanoseconds.

> But there is a specialized command that lets you specify all 3 inputs and get an updated duration.

is this available yet?

bobhy commented:
No additional features today, but the unit tests all work.
I don't know why typos is failing -- the errors are not in my changes!

bobhy commented:
Check progress vs handy checklist in [docs/README-duration-spec.md](https://github.com/bobhy/nushell/blob/duration/docs/README-duration-spec.md/).

Focusing on completing the arithmetic and comparison operations involving duration.

bobhy commented:
I don't claim to be taking the shortest path through this.  I'm taking the
"robust" approach of re-implementing the Duration type as a quantity+units
rather than implicitly scaled so I can handle the durations >= months.  And
*that* means a lot of work preserving the implicit coercion to int NS in
all the arithmetic operations.
But I think I can focus on the builtins now and implement the
duration-aware calculations that got me going down this path in the first
place.

re:
1. I really don't think we can keep the "humanized" rendering as the
default string representation of a duration value.  It's a *conversion*,
and, if it is to be accurate, it must know the starting date for the
duration in order to properly show days, weeks *and months*. That boundary
between weeks and months is where all the ugly is.  Nor do I feel we're
doing the user any service by offering an option to show inaccurate
results.  So my take on it is the stringification of the value is a simple
1:1 mapping from the internal value, and the humanized representation
requires user engagement.  At the very least, user should have a chance to
specify the format desired: should it be "M months, W weeks, D days" or
just "M months, D days"?  Should it be "S seconds, L milliseconds, U
microseconds, N nanoseconds" or "S.LLLUUUNNN seconds"?  I haven't
spec'ed it yet, but there should be an `list<duration> | into string
--format` which or such which formats the  normalized list of durations
into string and can handle these possibilities.  In the meantime, user can
roll his/her own.

2. You're right base date is inferred whenever there's a date term in the
arithmetic expression, so you can get accurate duration without having to
specify --base-date.  For builtins where there is no natural date input,
such as `duration | into duration --normalize`, the `--base-date` is only
required if the operation needs to span the weeks / months divide.

3. I think rendering of the duration ought to be themable, so user can
decide both the punctuation and the colorization.  I think we can cover all
the bases if the closure for duration theme is invoked with `{ | quan unit
| ... }`



On Mon, Jun 12, 2023 at 9:18 AM Darren Schroeder ***@***.***>
wrote:

> Maybe I was naive but I didn't think it would be this much work. Seems
> like you're working really hard and diligently on this. Thanks for this! I
> have a comment below that I want to bring up because I'd hate to get to the
> end and have significant changes after you've put in so much time and
> effort.
>
> The thing that I'm not quite understanding is part of this
>
> 〉let epoch = ((date now) - 1970-01-01T00:00:00z)
> 〉$epoch
> 1686077974017025614_nanoseconds
> 〉let norm = $epoch | into duration --normalize --base-date 1970-01-01T00:00:00z
> 〉$norm
> [53_years 5_months 5_days 20_hours 30_minutes 14_seconds 102_milliseconds 94_microseconds 230_nanoseconds]
>
> 〉$norm | each {into record | $"($in.quantity) ($in.unit)"} | str join ", "   # format 2
> 53 years, 5 months, 5 days, 20 hours, 30 minutes, 14 seconds, 102 milliseconds, 94 microseconds, 230 nanoseconds
>
>
>    1. Currently, all durations are represented in human readable
>    durations. I think we want to keep that as the default. We don't want to
>    have to specify --normalize and --base-date. Maybe we should have a "rough"
>    and "precise" mode like chrono-humanize
>    <https://gitlab.com/imp/chrono-humanize-rs/-/blob/master/src/humantime.rs>
>    2. It seems like base date could be inferred in some cases when it's
>    datetime - datetime or datetime + datetime, and maybe others like datetime
>    +/- duration?
>    3. When printing human readable durations, I think we should keep our
>    default abbreviations the same as they are now with the exception of maybe
>    adding the plural form when it applies.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nushell/nushell/pull/9113#issuecomment-1587328980>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPBIZA5IBXMXG5CGIJKEQTXK4JIRANCNFSM6AAAAAAXXT4IPA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

fdncred commented:
> I really don't think we can keep the "humanized" rendering as the
default string representation of a duration value.

This may be a deal-breaker, but I'll withhold judgement until we get down the road a bit further. I'd love to be wrong about it and I'd love to have something more accurate here. So, keep chugging along.

I also like your idea about theming duration, although I can't really envision how it would work exactly but it sounds cool.

bobhy commented:
I'm sorry to have gone silent on this PR for so long!
I've been taking a long look at the work remaining and am doubtful it will ever all get done!  I'm trying to improve too many things at once.  Time to simplify...

First step shall be to fix what's clearly broken with the current nushell duration type:
* to_string representation shows years and months assuming 30 day/ mo.  Nushell default banner is *always* wrong.
* it's very hard to do accounting-accurate datetime subtraction, with results in hours or days.
```
> 2022-03-01 - 2022-02-28
1day   # right
> 2020-03-01 - 2020-02-28
2day  # right
> 2022-03-01 - 2020-02-28
2yr 2day  # wrong: should be 2yr 1day
```
* cannot convert to_string form of duration back to a duration value.  Likewise for record --> duration.

I'm going to revert the core duration type changes in this PR and just fix the above, see whether that's good enough.

fdncred commented:
> I'm sorry to have gone silent on this PR for so long!

I'm all for you taking 1 bite at a time at this and think that it's a great idea. It'll make it easier to review too since there have been some discussions in the threads above about what we want vs what this PR does.

> Nushell default banner is _always_ wrong

It's pretty close these days thanks to the script JT wrote in the stdlib that uses the datetime-diff function. But I agree, without this script it's very off.

As a way of finding some ground truth, I've been using this website.

https://www.timeanddate.com/date/timezoneduration.html?d1=10&m1=05&y1=2019&d2=8&m2=7&y2=2023&h1=09&i1=59&s1=12&h2=08&i2=46&s2=06

Which should be `2023-07-08T08:46:06-05:00` minus `2019-05-10T09:59:12-07:00` which, according to the website, comes out to `4 years, 1 month, 27 days, 20 hours, 46 minutes, 54 seconds`.

Our datetime-diff says
```
❯ (datetime-diff ($now) 2019-05-10T09:59:12-07:00)
╭─────────────┬─────╮
│ year        │ 4   │
│ month       │ 1   │
│ day         │ 28  │
│ hour        │ 20  │
│ minute      │ 46  │
│ second      │ 54  │
│ nanosecond  │ 0   │
│ millisecond │ 807 │
│ microsecond │ 207 │
╰─────────────┴─────╯
```
which is very close, but 1 day off (wonder if there's a bug in the script), but anyway, it's much closer than it used to be.

bobhy commented:
Agree with all of the above.
I think the key observation is that duration type has limited range (no
months, years) and the tool that prints the banner output does not use
duration type (all by itself).  But I continue to think that the months and
years calculation is also an important scenario and we need a tool that
does it, but has a nice relationship with duration type.

On Sat, Jul 8, 2023 at 7:17 AM Darren Schroeder ***@***.***>
wrote:

> I'm sorry to have gone silent on this PR for so long!
>
> I'm all for you taking 1 bite at a time at this and think that it's a
> great idea. It'll make it easier to review too since there have been some
> discussions in the threads above about what we want vs what this PR does.
>
> Nushell default banner is *always* wrong
>
> It's pretty close these days thanks to the script JT wrote in the stdlib
> that uses the datetime-diff function. But I agree, without this script it's
> very off.
>
> As a way of finding some ground truth, I've been using this website.
>
>
> https://www.timeanddate.com/date/timezoneduration.html?d1=10&m1=05&y1=2019&d2=8&m2=7&y2=2023&h1=09&i1=59&s1=12&h2=08&i2=46&s2=06
>
> Which should be 2023-07-08T08:46:06-05:00 minus 2019-05-10T09:59:12-07:00
> which, according to the website, comes out to 4 years, 1 month, 27 days,
> 20 hours, 46 minutes, 54 seconds.
>
> Our datetime-diff says
>
> ❯ (datetime-diff ($now) 2019-05-10T09:59:12-07:00)
> ╭─────────────┬─────╮
> │ year        │ 4   │
> │ month       │ 1   │
> │ day         │ 28  │
> │ hour        │ 20  │
> │ minute      │ 46  │
> │ second      │ 54  │
> │ nanosecond  │ 0   │
> │ millisecond │ 807 │
> │ microsecond │ 207 │
> ╰─────────────┴─────╯
>
> which is very close, but 1 day off (wonder if there's a bug in the
> script), but anyway, it's much closer than it used to be.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nushell/nushell/pull/9113#issuecomment-1627330818>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPBIZHJCMEFVBPPFZI4U43XPFTXPANCNFSM6AAAAAAXXT4IPA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bobhy commented:
This PR has grown much too unwieldy and has some changes that are controversial.  It is being closed in favor of #9632, which takes a more limited approach.
